### PR TITLE
Added tests for LoginViewModel

### LINKED ISSUE #983: Create tests for the LoginViewModel class
### Chore

**Description**
Write tests for the LoginViewModel class.

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by sridharjajoo**
Fixes #983 
Changes: 

- A viewmodel package was created in which all test related to viewmodel will be written.
-  testImplementation dependencies were added for unit testing LiveData
- TracksPresenterTest.class had an incorrect test method name which was refactored.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -200,7 +201,13 @@ dependencies {
     androidTestImplementation('com.android.support.test.espresso:espresso-core:3.0.1', {
         exclude group: 'com.android.support', module: 'support-annotations'
     })
+    testImplementation("android.arch.core:core-testing:$versions.lifecycle", {
+        exclude group: 'com.android.support', module: 'support-compat'
+        exclude group: 'com.android.support', module: 'support-annotations'
+        exclude group: 'com.android.support', module: 'support-core-utils'
+    })
     testImplementation 'junit:junit:4.12'
+    testImplementation "android.arch.core:core-testing:$versions.lifecycle"
```
Comment (iamareebjamal) [inline app/build.gradle]: Why twice?

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
+package org.fossasia.openevent.app.core.viewmodel;
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Should be in the same package as the Login ViewModel

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+@RunWith(JUnit4.class)
+public class LoginViewModelTest {
+
+    @Rule public MockitoRule mockitoRule = MockitoJUnit.rule();
+    @Rule public TestRule rule = new InstantTaskExecutorRule();
+
+    @Mock private Preferences sharedPreferenceModel;
+    @Mock private AuthService authModel;
+    private final HostSelectionInterceptor interceptor = new HostSelectionInterceptor();
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Why not mocking

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        progress.onChanged(true);
+        verify(progress).onChanged(true);
+
+        login.onChanged(true);
+        verify(progress).onChanged(true);
+
+        progress.onChanged(false);
+        verify(progress).onChanged(false);
+    }
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: This test is not testing the ViewModel

This is testing itself

Call hello on speaker
Verify that hello was called on speaker

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        //given
+        String errorString = "Test Error!";
+        loginViewModel.getError().observeForever(error);
+
+        //when
+        error.onChanged(errorString);
+
+        //then
+        verify(error).onChanged(errorString);
+    }
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Same

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Test
+    public void shouldLoginAutomatically() {
+        loginViewModel.getLoginStatus().observeForever(login);
+        login.onChanged(true);
+        verify(login).onChanged(true);
+    }
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Same

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Test
+    public void shouldNotLoginAutomatically() {
+        loginViewModel.getLoginStatus().observeForever(login);
+        login.onChanged(false);
+        verify(login).onChanged(false);
+    }
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Same

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        loginViewModel.getEmailList().observeForever(email);
+        loginViewModel.getLoginStatus().observeForever(login);
+
+        //checking whether user is logged out
+        login.onChanged(false);
+        verify(login).onChanged(false);
+
+        email.onChanged(SAVED_EMAILS);
+        verify(email).onChanged(SAVED_EMAILS);
+     }
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Same

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        loginViewModel.getEmailList().observeForever(email);
+        loginViewModel.getLoginStatus().observeForever(login);
+
+        //checking whether user is logged in
+        login.onChanged(true);
+        verify(login).onChanged(true);
+
+        email.onChanged(sharedPreferenceModel.getStringSet(Constants.SHARED_PREFS_SAVED_EMAIL, null));
+        verify(email, Mockito.never()).onChanged(SAVED_EMAILS);
+    }
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Same

**Comment by sridharjajoo** [comment]
I have only changed the shouldShowLoginError test and rest others are same

File: app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java [hunk 10]
```diff
@@ -43,8 +42,8 @@ public LoginViewModel(AuthService loginModel, HostSelectionInterceptor intercept
 
     //for logging into the app
     public void login() {
+        progress.setValue(false);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java]: Why posting false?

File: app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java [hunk 11]
```diff
@@ -60,8 +62,8 @@ public void shouldLoginSuccessfully() {
         loginViewModel.getProgress().observeForever(progress);
         loginViewModel.getLoginStatus().observeForever(login);
 
-        login.onChanged(false);
-        verify(login).onChanged(false);
+        progress.onChanged(true);
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/viewmodel/LoginViewModelTest.java]: Why?

**Comment by iamareebjamal** [comment]
@sridharjajoo Again, you should see the implementation of previous tests. You are triggering and verifying it yourself

File: app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java (hunk 10 above)
Comment (sridharjajoo) [reply app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java]: So what change should I be doing here?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java]: true

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java]: Even so, you should be doing that through doOnSubscribe

**Comment by sridharjajoo** [comment]
@iamareebjamal could you see the shouldShowLoginError test. I have changed it from the previous commit.

**Comment by iamareebjamal** [comment]
@sridharjajoo What's the error there?

**Comment by sridharjajoo** [comment]
@iamareebjamal It also involves a call to login( ) . So facing NPE there. Apart from that just wanted to confirm that has it been implemented correctly?

**Comment by iamareebjamal** [comment]
It's implemented correctly but post the stack trace of NPE or debug. You could see that progress was null previously. See what is null in this case

File: app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java (hunk 10 above)
Comment (sridharjajoo) [reply app/src/main/java/org/fossasia/openevent/app/core/auth/login/LoginViewModel.java]: progress has already been set to true in doOnSubscribe

**Comment by sridharjajoo** [comment]
@iamareebjamal Any more changes required?

File: app/src/test/java/org/fossasia/openevent/app/core/presenter/LoginViewModelTest.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
+package org.fossasia.openevent.app.core.presenter;
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/presenter/LoginViewModelTest.java]: Wrong package. Change to as discussed on gitter

File: app/src/test/java/org/fossasia/openevent/app/core/presenter/LoginViewModelTest.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        Mockito.verify(email).onChanged(SAVED_EMAILS);
+    }
+
+    @Test
+    public void shouldNotAttachEmailAutomatically() {
+        Mockito.when(sharedPreferenceModel.getStringSet(Constants.SHARED_PREFS_SAVED_EMAIL, null)).thenReturn(null);
+        Mockito.when(authModel.isLoggedIn()).thenReturn(false);
+
+        loginViewModel.getEmailList().observeForever(email);
+        loginViewModel.getLoginStatus().observeForever(login);
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/presenter/LoginViewModelTest.java]: You haven't triggered vm function here

File: app/src/test/java/org/fossasia/openevent/app/core/presenter/LoginViewModelTest.java (hunk 12 above)
Comment (sridharjajoo) [reply app/src/test/java/org/fossasia/openevent/app/core/presenter/LoginViewModelTest.java]: Just to confirm, the package will be app.core.auth.login as it hs the vm

**Comment by sridharjajoo** [comment]
@iamareebjamal Can you approve again.I actually had to remove an unused import.

**Pull request merged**